doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1149 @@
+time.
+
+Once the datatype is found (a step that is skipped if the user provides it),
+the actual descriptor has to be discovered.
+For most datatypes, which are not parametric, this is always a canonical default
+instance and ``__discover_descr_from_pyobject__`` does not need to be defined.
+For parametric datatypes, however, such as strings, it is necessary to find the
+correct string length and thus inspect the given python object, this is
+also necessary for example for a Categorical datatype.
+In this case ``__discover_descr_from_pyobject__`` is called and must return
+a dtype instance.
+*Implementors should aim to avoid complex logic in this function.*
+
+After the dtype instance is found, the common dtype with the current previous
+found common dtype instance to find a single dtype that represents all inputs
+and thus support for example mixed integer and floating point input.
+
+Any object which cannot be associated with a dtype will be handled as an
+array-like or sequence and discovered recursively.
+
+If one of the objects is an array, or array-like object, its datatype is
+used directly.
+In the case where the user provided a DType, it will be force-cast to the
+user given datatype (class). Force-casting is the current behaviour, and this
+may be made more restrictive in the future.
+
+In the case that an object is neither a known type, an array-like, or a
+sequence.  The object will be regarded as handled as being of ``object`` dtype.
+
+**Limitations:**
+
+The above issue 3. is currently (sometimes) supported by NumPy so that
+the values of an included array are inspected.
+Support in those cases may be kept for compatibility, however,
+it will not be exposed to user datatypes.
+This means that if e.g. an array with a parametric string dtype is coerced above
+(or cast) to an array of a fixed length string dtype (with unknown length),
+this will result in an error.
+Such a conversion will require passing the correct DType (fixed length of the
+string) or providing a utility function to the user.
+
+The use of a global type map means that an error or warning has to be given
+if two DTypes wish to map to the same python type, in most cases DTypes user
+DTypes should only be implemented for types defined within the same library to
+avoid the potential for conflicts and .
+It is the DType implementors responsibility to be careful about this and use
+the flag to disable registration when in doubt.
+
+**Alternatives:**
+
+The above proposes to add a global mapping, however, initially limiting it
+to types deriving from a NumPy subclass (and a fixed set of Python types).
+This could be relaxed in the future.
+Alternatively, we could rely on the scalar belonging to the user dtype to
+implement ``scalar.dtype`` or similar.
+Initially, at the moment the exact implementation shall be *undefined*.
+Scalars will have to derive from a NumPy scalar, and provide at least a
+``.dtype`` attribute. Although, this may not be used within NumPy.
```

Comment:
```suggestion
Scalars will have to derive from a NumPy scalar and provide at least a
``.dtype`` attribute, though, this may not be used within NumPy.
```